### PR TITLE
Add hints for the case of confusing enum with its variants

### PR DESCRIPTION
**Pull request by Menschenkindlein**
A solution for https://github.com/rust-lang/rust/issues/43871. When one uses an enum in a place that accepts variants (e.g., `Option(result)` instead of `Some(result)`), suggest one of this enum's variants.

cc @estebank

### CODE REVIEW

File: src/librustc_resolve/lib.rs [hunk 1]
```diff
@@ -2599,6 +2599,21 @@ impl<'a> Resolver<'a> {
                         }
                         _ => {}
                     },
+                    (Def::Enum(..), PathSource::TupleStruct) => {
+                        if let Some(variants) = this.collect_enum_variants(def) {
+                            err.note(&format!("did you mean to use one \
+                                                of the following variants?\n{}",
+                                variants.iter()
+                                    .map(|suggestion| format!("- `{}`",
+                                                               path_names_to_string(suggestion)))
```
Comment (estebank) [inline src/librustc_resolve/lib.rs]: nitpick, move indentation one char to the left

File: src/librustc_resolve/lib.rs [hunk 2]
```diff
@@ -2599,6 +2599,21 @@ impl<'a> Resolver<'a> {
                         }
                         _ => {}
                     },
+                    (Def::Enum(..), PathSource::TupleStruct) => {
+                        if let Some(variants) = this.collect_enum_variants(def) {
+                            err.note(&format!("did you mean to use one \
+                                                of the following variants?\n{}",
```
Comment (estebank) [inline src/librustc_resolve/lib.rs]: nitpick, move indentation one char to the left

**Review by estebank — CHANGES_REQUESTED** [review]
It looks good to me, could you add a couple of ui tests?

**Comment by Menschenkindlein** [comment]
@estebank Could you, please, give me more guidance? I don't really know where is the place for the ui tests, and how to write them.

**Comment by oli-obk** [comment]
@Menschenkindlein you can add a new file to `src/test/ui` and follow the error message that you get when running `./x.py test src/test/ui`

**Comment by estebank** [comment]
@Menschenkindlein You can find a bit more of an explanation in [COMPILER_TESTS.md](https://github.com/rust-lang/rust/blob/master/src/test/COMPILER_TESTS.md#guide-to-the-ui-tests), but what you have to do is pretty much what @oli-obk explained. After you've ran `./x.py test src/test/ui`, it'll fail (because you don't have an `.stderr` file in the same place as your new test `.rs` file). After you run [`update-all-references.sh`](https://github.com/rust-lang/rust/blob/master/src/test/ui/update-all-references.sh) you'll have that `.stderr` file in place so subsequent runs of `./x.py test src/test/ui` should succeed.

**Comment by Menschenkindlein** [comment]
I added some tests. It looks like it doesn't work in tests. When I created a separate project and tested the feature, it worked. Somehow it manages to suggest an empty list of variants.

By the way, is it a reasonable idea to add a suggestion of a variant in other places where they can be used, e.g. when they are used as constructors?

**Comment by kennytm** [comment]
Hi @estebank, looks like @Menschenkindlein needs some help from you! 😃

**Comment by Menschenkindlein** [comment]
I assumed that one `Def` could designate more than one module, and now my code works. But this assumption looks very strange to me. What should I do?

**Comment by estebank** [comment]
@Menschenkindlein it is slightly odd if you're finding more than one module per `Def`, but the code so far looks good. Could you try adding some `debug!()` statements in `find_module` and run your local build with `RUST_LOG=debug ./stage1/bin/rustc file.rs` in order to verify that you're indeed ever receiving more than 1 `ImportSuggestion`?

**Comment by Menschenkindlein** [comment]
There indeed were two paths to `Option`. Probably, it was so because of reexports: the same module was found by going from two different paths. But the real bug was that I forgot to "populate" the module in `collect_enum_variants`. Continuing the search in `find_module` populated it and unintentionally solved the problem.
Now I added `self.populate_module_if_necessary(enum_module)` and it works!

**Comment by estebank** [comment]
Awesome @Menschenkindlein! I really like this feature!

r=me once you've fixed `ui/did_you_mean/issue-43871-enum-instead-of-variant.stderr`

```
[00:45:25]  error[E0423]: expected function, found enum `Option`
[00:45:25]    --> $DIR/issue-43871-enum-instead-of-variant.rs:14:13
[00:45:25]     |
[00:45:25]  14 |     let x = Option(1);
[00:45:25]     |             ^^^^^^
[00:45:25]     |
[00:45:25]     = note: did you mean to use one of the following variants?
[00:45:25] -           - `std::prelude::v1::Option::None`
[00:45:25] -           - `std::prelude::v1::Option::Some`
[00:45:25] +           - `std::option::Option::None`
[00:45:25] +           - `std::option::Option::Some`
[00:45:25]  
[00:45:25]  error[E0532]: expected tuple struct/variant, found enum `Option`
[00:45:25]    --> $DIR/issue-43871-enum-instead-of-variant.rs:16:12
[00:45:25]     |
[00:45:25]  16 |     if let Option(_) = x {
[00:45:25]     |            ^^^^^^
[00:45:25]     |
[00:45:25]     = note: did you mean to use one of the following variants?
[00:45:25] -           - `std::prelude::v1::Option::None`
[00:45:25] -           - `std::prelude::v1::Option::Some`
[00:45:25] +           - `std::option::Option::None`
[00:45:25] +           - `std::option::Option::Some`
[00:45:25]  
[00:45:25]  error[E0532]: expected tuple struct/variant, found enum `Example`
[00:45:25]    --> $DIR/issue-43871-enum-instead-of-variant.rs:22:12
[00:45:25]     |
[00:45:25]  22 |     if let Example(_) = y {
[00:45:25]     |            ^^^^^^^
[00:45:25]     |
[00:45:25]     = note: did you mean to use one of the following variants?
[00:45:25]             - `Example::Ex`
[00:45:25]             - `Example::NotEx`
[00:45:25]  
[00:45:25]  error: aborting due to 3 previous errors
[00:45:25]  
```

**Comment by oli-obk** [comment]
I don't think they are fixable. These are order issues depending on HashMap randomness.

You can turn them into compile-fail tests, since they don't care about order or the full message.

```
//~ SUGGESTION Option::Some
```

Should work

**Comment by kennytm** [comment]
If both modules can appear, you could add a [custom normalization rule](https://github.com/rust-lang/rust/blob/master/src/test/COMPILER_TESTS.md#normalization) that maps `std::prelude::v1::Option` to `std::option::Option` (as long as `None` always appear above `Some`).

```rust
// normalize-stderr: "std::prelude::v1::Option" -> "std::option::Option"
```

**Comment by Menschenkindlein** [comment]
I implemented `find_module` using `for_each_child`. `for_each_child_stable` should solve the issue with randomness.

**Comment by oli-obk** [comment]
I'm not sure if that solves it. This error is emitted during resolve, right? So we aren't fully resolved yet, thus the paths in find_module can't necessarily find the correct module. I had a similar issue and solved it by storing the information necessary to produce the diagnostic and emit the error after resolve is finished. There's several of these delayed diagnostics in resolve.

**Comment by Menschenkindlein** [comment]
@estebank Thank you for patiently guiding me to the completion of this pull request!

**Pull request merged**